examples/timeseries_classification_from_scratch.qmd

```diff
@@ -0,0 +1,173 @@
+---
+title: "Timeseries classification from scratch"
+authors:
+  - "[hfawaz](https://github.com/hfawaz/)"
+  - "[terrytangyuan](https://github.com/terrytangyuan) - R adaptation"
+  - "[t-kalinowski](https://github.com/t-kalinowski) - R adaptation"
+date-created: 2022/12/03
+date-last-modified: 2022/12/03
+description: "Training a timeseries classifier from scratch on the FordA dataset from the UCR/UEA archive."
+categories: [timeseries]
+aliases:
+  - ../guide/keras/examples/timeseries_classification_from_scratch/index.html
+---
+
+## Introduction
+
+This example shows how to do timeseries classification from scratch, starting from raw
+CSV timeseries files on disk. We demonstrate the workflow on the FordA dataset from the
+[UCR/UEA archive](https://www.cs.ucr.edu/%7Eeamonn/time_series_data_2018/).
+
+## Setup
+
+```{r}
+library(tensorflow)
+library(keras)
+set.seed(1234)
+```
+
+```{r, message=FALSE}
+train_df <- readr::read_tsv("https://raw.githubusercontent.com/hfawaz/cd-diagram/master/FordA/FordA_TRAIN.tsv", col_names = FALSE)
+x_train <- train_df[, -1]
+y_train <- train_df[, 1]
+
+test_df <- readr::read_tsv("https://raw.githubusercontent.com/hfawaz/cd-diagram/master/FordA/FordA_TEST.tsv", col_names = FALSE)
+x_test <- test_df[, -1]
+y_test <- test_df[, 1]
+```
+
+## Visualize the data
+
+Here we visualize one timeseries example for each class in the dataset.
+
+```{r}
+library(ggfortify)
+
+ts_obj <- ts(tibble::tibble("Class -1" = as.numeric(x_train[1, ]), "Class 1" = as.numeric(x_train[2, ])))
+autoplot(ts_obj, ts.geom = 'line', facets = FALSE)
+```
+
+## Standardize the data
+
+Our timeseries are already in a single length (500). However, their values are
+usually in various ranges. This is not ideal for a neural network;
+in general we should seek to make the input values normalized.
+For this specific dataset, the data is already z-normalized: each timeseries sample
+has a mean equal to zero and a standard deviation equal to one. This type of
+normalization is very common for timeseries classification problems, see
+[Bagnall et al. (2016)](https://link.springer.com/article/10.1007/s10618-016-0483-9).
+
+In order to use `sparse_categorical_crossentropy`, we will have to count
+the number of classes beforehand.
+
+```{r}
+num_classes <- nrow(unique(y_train))
```

Comment:
Calling `unique()` on a dataframe is usually bad practice (it's generally slow, since it casts each row as a string by printing it, and then comparing the printed representation). 

`unique` is great for atomic vectors however. I would change this to `length(unique(y_train[[1]]))` or `length(unique(train_df$y))`.